packages/x-data-grid/src/hooks/features/rows/useGridRowsMeta.ts

```diff
@@ -183,113 +182,93 @@ export const useGridRowsMeta = (
-    (id: GridRowId, height: number) => {
-      rowsHeightLookup.current[id].sizes.baseCenter = height;
-      rowsHeightLookup.current[id].isResized = true;
-      rowsHeightLookup.current[id].needsFirstMeasurement = false;
-      hydrateRowsMeta();
-    },
-    [hydrateRowsMeta],
-  );
+    isHeightMetaValid.current = true;
+  }, [apiRef, pinnedRows, currentPage.rows, processHeightEntry]);
 
-  const debouncedHydrateRowsMeta = React.useMemo(
-    () => debounce(hydrateRowsMeta, props.rowPositionsDebounceMs),
-    [hydrateRowsMeta, props.rowPositionsDebounceMs],
-  );
+  const getRowHeight: GridRowsMetaApi['unstable_getRowHeight'] = (rowId) => {
+    return heightCache.get(rowId)?.content ?? rowHeight;
+  };
 
-  const storeMeasuredRowHeight = React.useCallback<
-    GridRowsMetaApi['unstable_storeRowHeightMeasurement']
-  >(
-    (id, height) => {
-      if (!rowsHeightLookup.current[id] || !rowsHeightLookup.current[id].autoHeight) {
-        return;
-      }
+  const storeRowHeightMeasurement: GridRowsMetaApi['unstable_storeRowHeightMeasurement'] = (
+    id,
+    height,
+  ) => {
+    const entry = apiRef.current.getRowHeightEntry(id);
 
-      // Only trigger hydration if the value is different, otherwise we trigger a loop
-      const needsHydration = rowsHeightLookup.current[id].sizes.baseCenter !== height;
+    const didChange = entry.content !== height;
 
-      rowsHeightLookup.current[id].needsFirstMeasurement = false;
-      rowsHeightLookup.current[id].sizes.baseCenter = height;
+    entry.needsFirstMeasurement = false;
+    entry.content = height;
 
-      if (needsHydration) {
-        debouncedHydrateRowsMeta();
-      }
-    },
-    [debouncedHydrateRowsMeta],
-  );
+    isHeightMetaValid.current &&= !didChange;
+  };
 
-  const rowHasAutoHeight = React.useCallback<GridRowsMetaPrivateApi['rowHasAutoHeight']>((id) => {
-    return rowsHeightLookup.current[id]?.autoHeight || false;
-  }, []);
+  const rowHasAutoHeight: GridRowsMetaPrivateApi['rowHasAutoHeight'] = (id) => {
+    return heightCache.get(id)?.autoHeight ?? false;
+  };
 
-  const getLastMeasuredRowIndex = React.useCallback<
-    GridRowsMetaPrivateApi['getLastMeasuredRowIndex']
-  >(() => {
+  const getLastMeasuredRowIndex: GridRowsMetaPrivateApi['getLastMeasuredRowIndex'] = () => {
     return lastMeasuredRowIndex.current;
-  }, []);
+  };
 
-  const setLastMeasuredRowIndex = React.useCallback<
-    GridRowsMetaApi['unstable_setLastMeasuredRowIndex']
-  >((index) => {
+  const setLastMeasuredRowIndex: GridRowsMetaApi['unstable_setLastMeasuredRowIndex'] = (index) => {
     if (hasRowWithAutoHeight.current && index > lastMeasuredRowIndex.current) {
       lastMeasuredRowIndex.current = index;
     }
-  }, []);
+  };
 
-  const resetRowHeights = React.useCallback(() => {
-    rowsHeightLookup.current = {};
+  const resetRowHeights: GridRowsMetaApi['resetRowHeights'] = () => {
+    heightCache.clear();
```

Comment:
We used `Object.create(null)` when creating the lookup but not when resetting it here. Using `Map` for lookups avoids the issues with `{}` (prototype injection, and presence of Object methods), and is also more performant.